importlib_metadata/__init__.py

```diff
@@ -482,15 +488,47 @@ def zip_children(self):
         zip_path = zipp.Path(self.root)
         names = zip_path.root.namelist()
         self.joinpath = zip_path.joinpath
-
         return dict.fromkeys(child.split(posixpath.sep, 1)[0] for child in names)
 
-    def search(self, name):
-        return (
-            self.joinpath(child)
-            for child in self.children()
-            if name.matches(child, self.base)
-        )
+    def update_cache(self):
+        root = self.root or "."
+        try:
+            mtime = os.stat(root).st_mtime
+        except OSError:
+            self.infos.clear()
+            self.eggs.clear()
+            self.last_mtime = -1
+            return
+        if mtime == self.last_mtime:
+            return
+        self.infos.clear()
+        self.eggs.clear()
+        base_is_egg = self.base.endswith(".egg")
+        for child in self.children():
+            low = child.lower()
+            if low.endswith((".dist-info", ".egg-info")):
+                # rpartition is faster than splitext and suitable for this purpose.
+                name = low.rpartition(".")[0].partition("-")[0]
+                normalized = Prepared.normalize(name)
+                self.infos.setdefault(normalized, []).append(child)
+            elif base_is_egg and low == "egg-info":
+                name = self.base.rpartition(".")[0].partition("-")[0]
+                legacy_normalized = Prepared.legacy_normalize(name)
+                self.eggs.setdefault(legacy_normalized, []).append(child)
+        self.last_mtime = mtime
+
+    def search(self, prepared):
+        self.update_cache()
+        if prepared.name:
+            infos = self.infos.get(prepared.normalized, [])
+            yield from map(self.joinpath, infos)
+            eggs = self.eggs.get(prepared.legacy_normalized, [])
+            yield from map(self.joinpath, eggs)
+        else:
+            for infos in self.infos.values():
+                yield from map(self.joinpath, infos)
+            for eggs in self.eggs.values():
+                yield from map(self.joinpath, eggs)
 
 
 class Prepared:
```

Comment:
Previously, this class was doing a lot of the heavy lifting. With this change, it now only stores the name, normalized name, and legacy_normalized name. Moreover, the encapsulation this class once provided is violated now that FastPath is calling into the Prepared methods. I'm not sure it's worth keeping this abstraction.